ggplot/ggplot.py

```diff
@@ -122,417 +84,389 @@ class _empty(object):
-            other scales will get a discrete scale.
+    def draw(self):
         """
-        # scale_type is up to now unused
-        # TODO: what happens if we add a second color mapping?
-        # Currently the color mapping in the legend is overwritten.
-        # What does ggplot do in such a case?
-        if legend_type in self.legend:
-            pass
-            #msg = "Adding a secondary mapping of {0} is unsupported and no legend for this mapping is added.\n"
-            #sys.stderr.write(msg.format(str(legend_type)))
-        self.legend[legend_type] = legend_dict
-
-
-def _is_identity(x):
-    if x in colors.COLORS:
-        return True
-    elif x in shapes.SHAPES:
-        return True
-    elif isinstance(x, (float, int)):
-        return True
-    else:
-        return False
+        Draw the main plot onto the axes.
+
+        Return
+        ------
+        out : ggplot
+            ggplot object with two new properties
+                - axs
+                - fig
+        """
+        data, panel, plot = self.plot_build()
+        fig, axs = plt.subplots(plot.facet.nrow,
+                                plot.facet.ncol,
+                                sharex=False,
+                                sharey=False)
+
+        axs = np.atleast_2d(axs)
+        axs = [ax for row in axs for ax in row]
+        for ax in axs[len(panel.layout):]:
+            ax.axis('off')
+        axs = axs[:len(panel.layout)]
+        plot.axs = axs
+        plot.fig = fig
+
+        # ax - axes for a particular panel
+        # finfo - panel (facet) information from layout table
+        for ax, (_, finfo) in zip(axs, panel.layout.iterrows()):
+            panel_idx = finfo['PANEL'] - 1
+            xy_scales = {'x': panel.x_scales[finfo['SCALE_X'] - 1],
+                         'y': panel.y_scales[finfo['SCALE_Y'] - 1]}
+
+            # Plot all data for each layer
+            for zorder, (l, d) in enumerate(
+                    zip(plot.layers, data), start=1):
+                bool_idx = (d['PANEL'] == finfo['PANEL'])
+                l.plot(d[bool_idx], xy_scales, ax, zorder)
+
+            # xaxis & yaxis breaks and labels and stuff
+            set_breaks_and_labels(plot, panel.ranges[panel_idx],
+                                  finfo, ax)
+            plot.theme.post_plot_callback(ax)
+
+            # TODO: Need to find a better place for this
+            # theme_apply turns on the minor grid only to turn
+            # it off here!!!
+            if isinstance(xy_scales['x'], scale_discrete):
+                ax.grid(False, which='minor', axis='x')
+
+            if isinstance(xy_scales['y'], scale_discrete):
+                ax.grid(False, which='minor', axis='y')
+
+            # draw facet labels
+            if isinstance(plot.facet, (facet_grid, facet_wrap)):
+                draw_facet_label(plot, finfo, ax, fig)
+
+        apply_facet_spacing(plot)
+        return plot
+
+    def plot_build(self):
```

Comment:
Makes more sense, but for some of those functions I named them similar to ggplot2 since the code closely follows it. In case there is a bug, I want to make it easy to cross correlate the two code bases.